### DISCUSSION TITLE
object has no attribute 'refresh'

### DISCUSSION THREAD

**Discussion by DdosTheCoder** [beeware/toga#2330] [Q&A] [answered]
I'm new to programming so sorry if this is actually a very easy and stupid question.

I'm trying to develop an Android training app for my friend with, because it's android, I can't create new windows, so I'm using `main_window.content`, and when I try to ran the app with `briefcase dev` I get an error in the cmd `AttributeError: 'Picture' object has no attribute 'refresh'`. I understand that's the error is in the fact, that the `Picture` class has no attribute `refresh` but I don't know how to fix it.

 I tried:
 
```
def refresh(widget): 
pass
```

and 

```
class Picture(toga.App, refresh):
```

but it doesn't work, I also looked in discussions and asked ChatGPT but nothing seems to work, so I came here.

This is a part of the code where I think the problem is. If you need I can send a full code here.

```
class Picture(toga.App):
    def startup(self):

        def barbell_bench_press(widget):

            barbell_bench_press_box = toga.Box()
            main_window.content = barbell_bench_press_box

           
            image_path =  toga.Image(self.paths.data / "C:\\Users\\ddosu\\Downloads\\barbell-bench-press.png")
            view = toga.ImageView(image_path, style=Pack(width=300, height=300))
            barbell_bench_press_box.add(view)
```

**Comment by rmartin16** [comment]
Please include what it is you are trying to do; it's difficult to provide advice if you don't state your relevant goals.

As far as the error is concerned, are you trying to call `Picture.refresh()` somewhere? If so, you'll need to define a method called `refresh` in your `Picture` class.

If you are able to provide a **minimal** app that reproduces the error (that only includes the code necessary to create your error), that can help since we'll have a straightforward example to commonly work from.

Also, please include the full error traceback as well; this is simplest if you just upload the log created by Briefcase. You can force this log creation with `briefcase dev --log`.

**Reply by DdosTheCoder** [reply]
I'm sorry for the confusion. 

My goal is: add picture to the box of the exercise , my code worked just fine before I tried to add a picture.

I don't know if you need it , but here's **minimal** app that shows how it's **supposed** to work:
```python
"""
S training - created to help
"""

import toga
from toga.style.pack import COLUMN, ROW, RIGHT, LEFT, Pack, CENTER



def startup(self):

    def barbell_bench_press(widget):

        barbell_bench_press_box = toga.Box()
        main_window.content = barbell_bench_press_box

        label = toga.Label(text="PICTURE OF THE BENCH PRESS")

        barbell_bench_press_box.add(label)


    push_box = toga.Box(style=Pack(direction=COLUMN))
    def push_list(widget):
        main_window.content = push_box

        barbell_bench_press_button = toga.Button("Barbell Bench Press", on_press=barbell_bench_press, style=Pack(flex=1, padding=0))

        push_box.add(barbell_bench_press_button, return_button)


    def return_to_main_window(widget):
        main_window.content = main_box


    main_window = toga.MainWindow(title="S training")

    toga.App.main_window = main_window
    main_window.show()


    split_box = toga.Box(style=Pack(direction=COLUMN, alignment="center"))

    dummy_box = toga.Box(style=Pack(flex=1))
    dummy_box_bottom = toga.Box(style=Pack(flex=1))

    push_button = toga.Button("Push", on_press=push_list, style=Pack(flex=1, padding=5))


    return_button = toga.Button("Go Back", on_press=return_to_main_window, style=Pack(padding_top=100))


    split_box.add(push_button)


    main_box = toga.Box(style=Pack(direction=COLUMN))
    main_box.add(dummy_box)
    main_box.add(split_box)
    main_box.add(dummy_box_bottom)

    main_window.content = main_box

    return main_box


def main():
    return toga.App("S training", "org.s_training", startup=startup, version="0.69.0")


if __name__ == "__main__":
    main().main_loop()

```





Here is the most minimalistic version of my code that I can think of:
```python
"""
S training - created to help
"""

import toga
from toga.style.pack import COLUMN, ROW, RIGHT, LEFT, Pack, CENTER


class Picture(toga.App):
    def startup(self):

        def barbell_bench_press(widget):

            barbell_bench_press_box = toga.Box()
            main_window.content = barbell_bench_press_box

            image_path =  toga.Image(self.paths.data / "C:\\Users\\ddosu\\Downloads\\barbell-bench-press.png")
            view = toga.ImageView(image_path, style=Pack(width=300, height=300))
            barbell_bench_press_box.add(view)


        push_box = toga.Box(style=Pack(direction=COLUMN))
        def push_list(widget):
            main_window.content = push_box

            barbell_bench_press_button = toga.Button("Barbell Bench Press", on_press=barbell_bench_press, style=Pack(flex=1, padding=0))

            push_box.add(barbell_bench_press_button, return_button)


        def return_to_main_window(widget):
            main_window.content = main_box


        main_window = toga.MainWindow(title="S training")

        toga.App.main_window = main_window
        main_window.show()


        split_box = toga.Box(style=Pack(direction=COLUMN, alignment="center"))

        dummy_box = toga.Box(style=Pack(flex=1))
        dummy_box_bottom = toga.Box(style=Pack(flex=1))

        push_button = toga.Button("Push", on_press=push_list, style=Pack(flex=1, padding=5))


        return_button = toga.Button("Go Back", on_press=return_to_main_window, style=Pack(padding_top=100))


        split_box.add(push_button)


        main_box = toga.Box(style=Pack(direction=COLUMN))
        main_box.add(dummy_box)
        main_box.add(split_box)
        main_box.add(dummy_box_bottom)

        main_window.content = main_box

        return main_box


def main():
    return toga.App("S training", "org.s_training", startup=Picture, version="0.69.0")


if __name__ == "__main__":
    main().main_loop()



```

And here is the logs: 
[briefcase.2024_01_06-21_21_40.dev.log](https://github.com/beeware/toga/files/13851313/briefcase.2024_01_06-21_21_40.dev.log)

**Reply by rmartin16** [reply, marked as answer]
Thanks.

In `main()`, you're returning a `toga.App` with `startup=Picture` but that's probably not going to work since invoking the object from `startup` needs to return a widget.

Try this instead:
```python
def main():
    return Picture()
```

You'll also want to consider the use of `main_box` in `return_to_main_window()` since it won't be available in that function.

**Reply by DdosTheCoder** [reply]
Thank you for fast response, I did like you told me, and ```AttributeError: 'Picture' object has no attribute 'refresh'``` error disappeared, but now I encounter another error, ```TypeError: App.main_loop() missing 1 required positional argument: 'self'```

I tried:
```
if __name__ == "__main__":
    main().main_loop(self)
```
    
but the error hasn't disappeared

Here are the logs:
[briefcase.2024_01_06-22_00_01.dev.log](https://github.com/beeware/toga/files/13851382/briefcase.2024_01_06-22_00_01.dev.log)

**Reply by DdosTheCoder** [reply]
I am very sorry it was the problem on my side, sorry that I disturbed you

**Reply by rmartin16** [reply]
Are you returning `Picture` or `Picture()` from `main()` in `app.py`? It needs to be `Picture()`.

**Reply by DdosTheCoder** [reply]
yes, the problem was in my bad typing, I reffered to Picture(), as Picture, thank you very much for your help and patience
